lib/cloud_controller/dependency_locator.rb

```diff
@@ -73,7 +75,7 @@ def copilot_client
     end
 
     def bbs_apps_client
-      @dependencies[:bbs_apps_client] || register(:bbs_apps_client, build_bbs_apps_client)
+      @dependencies[:bbs_apps_client] || register(:bbs_apps_client, build_apps_client)
```

Comment:
can you give us more details as to why you chose the dependency locator as the layer at which the bbs/opi swap occurs?

Reply:
For example, we were wondering if some type of [backend](https://github.com/cloudfoundry/cloud_controller_ng/tree/master/lib/cloud_controller/backends) adapter would work better? Or a "bridge component"-esque proxy external to CC that would convert BBS requests to OPI/Eirini requests?

Not really prescribing anything here, just want to have a discussion around the design.

Reply:
We wanted to be as least intrusive as possible and to make the minimum amount of changes needed to get OPI running from the CloudController side. Since most of the scheduler code in CC already works for us, we decided to reuse as much of it as possible, which meant that we only had to change the clients that directly communicate with the scheduler. It seems to us that if we create our own backend adapter we’ll have to re-implement a huge chunk of the Diego code that is already doing what we need.

Regarding the bridge-component idea, that would mean we’d have to convert the Diego recipes to OPI native requests. That approach would closely couple us with the Diego implementation, which is not what our goal is. Additionally, as a result of creating our OPI clients, we had to move the BBS recipe building logic (which is a BBS implementation detail) in the BBS clients, which creates a cleaner abstraction on this level in our opinion.

Note that we’re not that familiar with the CC code and we’re always open for a discussion and ideas. 🙂